Scheduled Confiugurations

amazor opened the pull request:
Phase two first implementation is complete, but needs debugging since ACL table is not being configured as expected.

This phase implements the ability to add to the CONFIG_DB a TIME_RANGE table. Once this configuration is sent, the schedulermgrd will pick up the changes, and call schedulermgr. This process will parse the interval, and create cron files under /etc/cron.d/ within the swss container. It will also create a default TIME_RANGE_STATUS_TABLE entry with the time range name, and give it the default field-value of "status": disabled.

To configure you may use this template json file:
  `{
          "TIME_RANGE": {
                  "rangeTest": {
                          "start": "*/2 * * * *",
                          "end": "1-59/2 * * * *"
                  },
                  "runOnceTest": {
                          "start": "20 * * * *",
                          "end": "21 * * * *",
                          "runOnce": "true"
                  }
          }
  }
  `
Then call this command on the json file: `sonic-cfggen -j <json-file-name> --write-to-db`.
This will trigger the code found in this PR.
Note this PR also needs changes from two other PRs in other repos. Since this is just phase 1, i decided to wait until both phases are complete before creating all the other PRs

amazor committed:
First Commit

created schedulermgrd and schedulermgr files

amazor committed:
Fixed bugs and Added Schedulermgr(d) to Makefile

Fixed minor and return task_process_status
Added systemd infrastructure functions to reload and update
Removed options for schedulermgrd application

amazor committed:
Remove systemd and use cron instead

amazor committed:
Fixed Bugs and Ran Code Formatter
- Fixed bug where endtime was used for enable cronjob
- Fixed bug where EOL was not added to crontab file, causing cron to not be able to read the file
- Removed extra c in writeContabFile() funcname
- Removed unnecessary scheduledConfigurationTable in schedulermgr
- Changed type of  m_stateTimeRangeStatusTable from ProducerStateTable to Table in order for set command to work correctly
- Formatted code

amazor committed:
Missed ending single quote for runOnce field

Forgot an ending quote when adding a runOnce==true boolean

aviramd commented on cfgmgr/schedulermgr.cpp:
```diff
@@ -0,0 +1,189 @@
+        return task_process_status::task_need_retry;
+    }
+
+    SWSS_LOG_INFO("Succesfully created crontab files for %s", taskName.c_str());
```
This log already exists inside  writeCrontabFile

amazor replied:
The log in writeCrontabFile is DEBUG/ERROR logs, this one is an INFO log. Also they are for a single file, as opposed to this log which specified both files have been successfully created.

matiAlfaro commented on cfgmgr/schedulermgrd.cpp:
```diff
@@ -0,0 +1,72 @@
+        auto schedulermgr = cfgOrchList[0];
+
+        swss::Select s;
+        for (Orch *o : cfgOrchList)
```
Why is this a vector? isn't there just one of these SchedulerMgr?

amazor replied:
You are correct that I can implement this in a more straightforward way without using a vector. But i copied how other *mgrd.cpp (macsecmgrd.cpp, intfmgrd.cpp, portmgrd.cpp, etc...) files look, even if there is only one object in the vector, they still add it to a vector.

I prefer to keep the style the same as other files.

Ex: 

        vector<Orch *> cfgOrchList = {&portmgr};

        swss::Select s;
        for (Orch *o : cfgOrchList)
        {
            s.addSelectables(o->getSelectables());
        }

matiAlfaro replied:
ok

matiAlfaro commented on cfgmgr/schedulermgr.cpp:
```diff
@@ -0,0 +1,189 @@
+    }
+    crontabFile << endl;
+    crontabFile.close();
+
```
General,
using && requires success for previous command executed, so we may get stuck with cron files that no one will ever delete

amazor replied:
Will replace with ';' instead to occur as a separate command. Also for deletion from CONFIG_DB in line 72

matiAlfaro commented on cfgmgr/schedulermgr.cpp:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    // Add time range status to range status table in state db
+    m_stateTimeRangeStatusTable.set(key, stateTableFieldValues);
```
What's the purpose of this table? Info in table is not very useful.

amazor replied:
This is the TIME_RANGE_STATUS_TABLE. This is the same table that the cronjobs will update. I want the time range to appear in the STATE_DB immediately, even if it is not enabled. This will most likely be used as a source for CLI show commands.

matiAlfaro commented on cfgmgr/schedulermgrd.cpp:
```diff
@@ -0,0 +1,72 @@
+            }
+
+            auto *c = (Executor *)sel;
+            c->execute();
```
what is this?

amazor replied:
I believe that this function executes the doTask(Consumer &consumer) function defined in schedulermgr.cpp. This is the main logic that occurs. The Select::TIMEOUT section (line numbers 57-61) call doTask() without any parameter which only clears any tasks that are left.

amazor committed:
Replace "&&" to ";" for cronjobs

Switching to ";" allows the configuration to be deleted after a "runOnce" whether or not the configuration changes succeed or not.

amazor committed:
Create timebasedconfigmgrd and rename schedulermgrd to timerangemgrd

amazor committed:
Rename timebasedconfigmgrd to scheduledconfigmgrd

matiAlfaro commented on cfgmgr/scheduledconfigmgr.cpp:
```diff
@@ -0,0 +1,403 @@
+        {
+            // Removing quotes from primitive types to match the expected format
+            string value = item.value().dump();
+            if (value.front() == '"' && value.back() == '"')
```
extract to utility function, reuse in join

matiAlfaro commented on cfgmgr/timerangemgr.cpp:
same as schedulermgr.cpp, which one do you want to review?

amazor replied:
timerangemgr.cpp is the current name. Need to remove schedulermgr.cpp.

matiAlfaro commented on cfgmgr/timerangemgr.cpp:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    // Create systemd files for time range and enable them
+    // TODO sanitize inputs
```
yes, verify time ranges validity.

amazor replied:
Need to strictly validate and sanitize time range since this can be a security vulnerability since we are running cron commands as root (albeit only in swss container).

matiAlfaro commented on cfgmgr/timerangemgr.cpp:
```diff
@@ -0,0 +1,189 @@
+    // TODO sanitize inputs
+    if (task_process_status::task_need_retry == createCronjobs(rangeName, start, end, (runOnce == "true")))
+    {
+        return task_process_status::task_need_retry;
```
what if succeeded in creating one file but failed on the second?

amazor replied:
Will implement a rollback mechanism in case of OS and possibly a validation errors.

matiAlfaro commented on cfgmgr/scheduledconfigmgrd.cpp:
```diff
@@ -0,0 +1,76 @@
+using namespace swss;
+
+/* SELECT() function timeout retry time, in millisecond */
+#define SELECT_TIMEOUT 1000
```
why do we need to "wake up" every second?

amazor replied:
The timeout will end up calling the doTask() function again with all the consumables still in the queue. This is usually used for "retrying" tasks that did not succeed the first time, but select will not return because no changes have been made to the DB to trigger it again.

matiAlfaro commented on cfgmgr/timerangemgrd.cpp:
I was expecting somewhere in this file to "receive" events from cron, if not - how does enable/disable (active/inactive) status is updated? by polling?

amazor replied:
timerangemgrd does not receive events from cron, it writes the cron files. It receives the time ranges from the user, and cron will activate/deactivate the ranges by writing to the STATE_DB. Scheduledconfigmgrd then receives an event from redis that a change has occured.

amazor committed:
Add support for removing configuration when inactive

Also clean up code, and began validation of cron syntax

amazor committed:
Added Multiple Updates (NOT TESTED)

- Added 3 new files to help with cron validation & time checking. Leaning to use croncpp.h and removing cronhelper.cpp ands cronhelper.h
- Added check that scheduled configuration does not already exist when adding a scheduled configuration
- Automatically apply scheduled configuration if time range is currently active
- Save scheduled configurations in "unboundConfiguration" hashmap when time range is deleted
- Check if scheduled configurations saved in "unboundConfiguration" hashmap should be bound to newly created time range, and then add it to that new time range if applicable
- Added if statement to not process time range/scheduled configuration if an invalid field is found in the asssociated table
- Implemented DEL operation processing for scheduled configuration and time range deletion
- New function "isTimeRangeActive()" in scheduledconfigmgr.cpp to check TIME_RANGE_STATUS table if the time range is active according to the field
- New function "is_time_in_range()" that uses the croncpp.h file in order to test if the local time is within the scheduled time range interval -- NOT TESTED
- Implemented DEL operation for when time range key is deleted from CONFIG_DB

matiAlfaro commented on cfgmgr/croncpp.h:
Wouldn't it be better to have the implementations in .cpp file

matiAlfaro commented on cfgmgr/croncpp.h:
```diff
@@ -0,0 +1,982 @@
+      static const cron_int CRON_MAX_YEARS_DIFF = 4;
+
+#ifdef CRONCPP_IS_CPP17
+      static const inline std::vector<std::string> DAYS = { "SUN", "MON", "TUE", "WED", "THU", "FRI", "SAT" };
```
Wouldn't it be easier (maybe better) to define DAYS and MONTHS as enums have have replace_ordinals be a bit smarter? What is the benefit of doing it this way?
also, these (and a few more) types/functions can be defined in a "more global" scope, so you won't have to define them twice

matiAlfaro commented on cfgmgr/croncpp.h:
```diff
@@ -0,0 +1,982 @@
+      inline std::tm to_tm(CRONCPP_STRING_VIEW time)
+      {
+         std::tm result;
+#if __cplusplus > 201103L
```
What is the purpose of these compiler version checks? isn't it simpler to write to the currently supported compiler

matiAlfaro commented on cfgmgr/timerangemgr.cpp:
```diff
@@ -8,15 +8,31 @@
+bool is_time_in_range(const cronexpr& startExpr, const cronexpr& endExpr, const std::tm& currentTM) {
+    std::time_t currentTime = mktime(const_cast<tm*>(&currentTM)); // Convert currentTM to time_t
+    time_t previousMinute = currentTime - 60; // Go back one minute to ensure full coverage
+
```
Doesn't this add risk of getting an "event" for something that you already activate here? do you handle such case?

amazor committed:
Implemented isTimeInRange() and removed changes added to croncpp.h

amazor committed:
Fixed Most Compiler Errors (Still Does Not Compile)

- Added license to croncpp.h
- Removed static qualifier from replace_ordinals() in croncpp.h
  - Done in order to remove warning saying this function isn't used
- change "enabled" and "disabled" to "active" and "inactive"
respectively

amazor committed:
Changed Data Structure ScheduledConfig DB (Compiles - not tested)

- Changed DB to use a nested unordered_map instead of a vector of pairs
- Surrounded replace_ordinals() in croncpp.h to remove unused fxn warning
- Surrounded make_cron() usage with try-catch

amazor committed:
Finished Basic Functionality of Feature

Will now correctly add and remove configurations when time range becomes active/inactive.
Time range status will be updated to active immediately if the current time is within the newly set time range.
Added internal database to keep track of active/inactive scheduled configurations in case of delete.

amazor committed:
Implemnted Year Functionality and Removed Seconds from croncpp.h

amazor committed:
Remove Unecessary Files
